File: src/api/categorias.ts
import { callCommand } from './client'
import type { Audit, ListQuery, LookupItem, PagedResult, RowVersion, Uuid } from './types'

/** See `docs/11-contratos-tauri.md` §5.12. */

export interface CategoriaFiltro {
  /** Matched against the name and the description, ignoring case. */
  texto?: string
  categoriaPadreId?: Uuid
  /** Root categories only. Distinct from not filtering by parent at all. */
  soloRaiz?: boolean
}

export interface CategoriaInput {
  nombre: string
  descripcion: string | null
  /** `#RRGGBB`; the backend stores it upper case. */
  colorHex: string | null
  icono: string | null
  categoriaPadreId: Uuid | null
}

export interface CategoriaListItem {
  id: Uuid
  nombre: string
  descripcion: string | null
  colorHex: string | null
  icono: string | null
  categoriaPadreId: Uuid | null
  categoriaPadreNombre: string | null
  movimientosCount: number
  hijasCount: number
  /** False when the category has movements or children, so the action can be disabled up front. */
  puedeEliminarse: boolean
  rowVersion: RowVersion
}

export interface CategoriaDetalle {
  id: Uuid
  nombre: string
  descripcion: string | null
  colorHex: string | null
  icono: string | null
  categoriaPadreId: Uuid | null
  movimientosCount: number
  hijasCount: number
  puedeEliminarse: boolean
  audit: Audit
}

/** Columns the backend accepts; anything else is rejected as a validation error. */
export const CATEGORIAS_SORTABLE = [
  'nombre',
  'movimientosCount',
  'hijasCount',
  'createdAt',
] as const

export function listCategorias(
  query: ListQuery<CategoriaFiltro>,
): Promise<PagedResult<CategoriaListItem>> {
  return callCommand('categorias_list', { query })
}

export function getCategoria(id: Uuid): Promise<CategoriaDetalle> {
  return callCommand('categorias_get', { id })
}

export function createCategoria(dto: CategoriaInput): Promise<CategoriaDetalle> {
  return callCommand('categorias_create', { dto })
}

export function updateCategoria(
  id: Uuid,
  dto: CategoriaInput,
  rowVersion: RowVersion,
): Promise<CategoriaDetalle> {
  return callCommand('categorias_update', { id, dto, rowVersion })
}

export function deleteCategoria(id: Uuid, rowVersion: RowVersion): Promise<void> {
  return callCommand('categorias_delete', { id, rowVersion })
}

export function lookupCategorias(texto?: string, limite?: number): Promise<LookupItem[]> {
  return callCommand('categorias_lookup', { texto, limite })
}

File: src-tauri/src/commands/categorias.rs
//! Commands of `categorias`. See `docs/11-contratos-tauri.md` §5.12.

use eo_application::dtos::categorias::{
    CategoriaDetalle, CategoriaFiltroDto, CategoriaInput, CategoriaListItem,
};
use eo_application::dtos::common::{ListQuery, LookupItem};
use eo_application::PagedResult;
use tauri::State;
use uuid::Uuid;

use crate::error::{handle, ApiResult};
use crate::state::AppState;

#[tauri::command]
pub async fn categorias_list(
    state: State<'_, AppState>,
    query: ListQuery<CategoriaFiltroDto>,
) -> ApiResult<PagedResult<CategoriaListItem>> {
    let outcome = match state.services() {
        Ok(services) => services.categorias.list(query).await,
        Err(e) => Err(e),
    };
    handle("categorias_list", outcome)
}

#[tauri::command]
pub async fn categorias_get(state: State<'_, AppState>, id: Uuid) -> ApiResult<CategoriaDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.categorias.get(id).await,
        Err(e) => Err(e),
    };
    handle("categorias_get", outcome)
}

#[tauri::command]
pub async fn categorias_create(
    state: State<'_, AppState>,
    dto: CategoriaInput,
) -> ApiResult<CategoriaDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.categorias.create(dto).await,
        Err(e) => Err(e),
    };
    handle("categorias_create", outcome)
}

#[tauri::command]
pub async fn categorias_update(
    state: State<'_, AppState>,
    id: Uuid,
    dto: CategoriaInput,
    row_version: String,
) -> ApiResult<CategoriaDetalle> {
    let outcome = match state.services() {
        Ok(services) => services.categorias.update(id, dto, &row_version).await,
        Err(e) => Err(e),
    };
    handle("categorias_update", outcome)
}

#[tauri::command]
pub async fn categorias_delete(
    state: State<'_, AppState>,
    id: Uuid,
    row_version: String,
) -> ApiResult<()> {
    let outcome = match state.services() {
        Ok(services) => services.categorias.delete(id, &row_version).await,
        Err(e) => Err(e),
    };
    handle("categorias_delete", outcome)
}

#[tauri::command]
pub async fn categorias_lookup(
    state: State<'_, AppState>,
    texto: Option<String>,
    limite: Option<u64>,
) -> ApiResult<Vec<LookupItem>> {
    let outcome = match state.services() {
        Ok(services) => services.categorias.lookup(texto, limite).await,
        Err(e) => Err(e),
    };
    handle("categorias_lookup", outcome)
}
